### PR TITLE
Snips dataset

### PR DESCRIPTION
**Pull request by alexeyr**
Add reader and (intent and NER) iterators for the https://github.com/snipsco/nlu-benchmark dataset

### CODE REVIEW

File: deeppavlov/dataset_readers/snips_reader.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            }
+
+        Args:
+            data_path: A path to a folder with dataset files.
+            queries_per_intent: Number of queries to load for each intent. None to load all.
+                If the requested number is greater than available in file, all queries are returned.
+            test_validate_split: Proportion of `_validate` files to be used as test dataset (since Snips
+                is split into training and validation sets without a separate test set).
+        """
+        # TODO download the files automatically
```
Comment (yoptar) [inline deeppavlov/dataset_readers/snips_reader.py]: If the files are not downloaded automatically, there should be an error or a warning if the data is not there

File: deeppavlov/dataset_readers/snips_reader.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+# Copyright 2019 Neural Networks and Deep Learning lab, MIPT
```
Comment (yoptar) [inline deeppavlov/dataset_readers/snips_reader.py]: Shouldn't this copyright be yours? :slightly_smiling_face:

File: deeppavlov/dataset_readers/snips_reader.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+from typing import List, Dict, Any, Optional
+from pathlib import Path
+
+from overrides import overrides
+
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.data.dataset_reader import DatasetReader
+
+
+@register('snips_reader')
```
Comment (yoptar) [inline deeppavlov/dataset_readers/snips_reader.py]: Registered classes should be added to the [registry.json](https://github.com/Huawei-MRC-OSI/DeepPavlov/blob/snips_dataset/deeppavlov/core/common/registry.json) file by hand or using the [utils.prepare.registry](https://github.com/Huawei-MRC-OSI/DeepPavlov/blob/snips_dataset/utils/prepare/registry.py) script that will, however, require all the local configs requirements

File: deeppavlov/dataset_readers/snips_reader.py (hunk 1 above)
Comment (alexeyr) [reply deeppavlov/dataset_readers/snips_reader.py]: I'll just put the expected intents list in, then an error should be raised from trying to read a non-existent file.

**Comment by alexeyr** [comment]
@dilyararimovna Sorry, I don't understand what you mean here.

File: deeppavlov/dataset_readers/snips_reader.py (hunk 3 above)
Comment (alexeyr) [reply deeppavlov/dataset_readers/snips_reader.py]: Fixed.

File: deeppavlov/dataset_readers/snips_reader.py (hunk 2 above)
Comment (alexeyr) [reply deeppavlov/dataset_readers/snips_reader.py]: Fixed :)

**Comment by yoptar** [comment]
About what @dilyararimovna said: there are configs for snips in [classifiers directory](https://github.com/Huawei-MRC-OSI/DeepPavlov/tree/snips_dataset/deeppavlov/configs/classifiers).
They use `basic_classification_reader` but could use the new one

**Comment by alexeyr** [comment]
@yoptar @dilyararimovna I see. For this, the original snips dataset needs to be uploaded to files.deeppavlov.ai first, I think.

**Comment by yoptar** [comment]
@alexeyr is http://files.deeppavlov.ai/datasets/snips.tar.gz ok?

**Comment by alexeyr** [comment]
@yoptar Yes, though `*_metrics.json` and `Snips_metrics_full.json` files can be removed. I'll update the request with configs (and automatic download) later then.

**Comment by alexeyr** [comment]
@yoptar Added downloading and updated configs.

File: deeppavlov/configs/evolution/evolve_intents_snips.json [hunk 4]
```diff
@@ -1,28 +1,13 @@
...
-    "split_fields": [
-      "train",
-      "valid"
-    ],
-    "split_proportions": [
-      0.9,
-      0.1
-    ]
+    "class_name": "snips_intents_iterator",
+    "seed": 42
```
Comment (yoptar) [inline deeppavlov/configs/evolution/evolve_intents_snips.json]:
```suggestion
    "seed": {
      "evolve_range": [
        50,
        500
      ],
      "discrete": true
    }
```

File: deeppavlov/dataset_readers/snips_reader.py (hunk 1 above)
Comment (yoptar) [reply deeppavlov/dataset_readers/snips_reader.py]: This `TODO` can be removed, it seems.

**Comment by alexeyr** [comment]
@yoptar Done.

**Comment by alexeyr** [comment]
@yoptar Note that I rebased to combine the commits, and this apparently happened between you approving the changes and merging.

**Pull request merged**